wcomponents-core/src/main/java/com/github/bordertech/wcomponents/WFileWidget.java

```diff
@@ -110,13 +133,109 @@ protected boolean doHandleRequest(final Request request) {
 		boolean changed = value != null || current != null;
 
 		if (changed) {
-			setData(value);
-		}
+			// Reset validation fields
+			resetValidationState();
+			
+			// if fileType is supplied then validate it
+			boolean validFileType;
+			if (isSetFileTypes()) {
+				validFileType = FileValidationUtil.validateFileType(value, getFileTypes());
+				addFileTypeValidation(validFileType, InternalMessages.DEFAULT_VALIDATION_ERROR_FILE_WRONG_TYPE, 
+						getFileTypes().toArray(new Object[getFileTypes().size()]));
+			} else {
+				validFileType = true;
+			}
+			
+			// if fileSize is supplied then validate it
+			boolean validFileSize;
+			if (isSetFileSize()) {
+				validFileSize = FileValidationUtil.validateFileSize(value, getMaxFileSize());
+				addFileSizeValidation(validFileSize, InternalMessages.DEFAULT_VALIDATION_ERROR_FILE_WRONG_SIZE, 
+						Util.readableFileSize(value.getSize()), Util.readableFileSize(getMaxFileSize()));
+			} else {
+				validFileSize = true;
+			}
+			
+			// if file is valid, the update data
+			if (validFileSize && validFileType) {
+				setData(value);
+			} else {
+				// otherwise no change
+				changed = false;
+			}
+		} 
 
 		return changed;
 	}
 
 	/**
+	 * Reset validation state.
+	 */
+	public void resetValidationState() {
+		getOrCreateComponentModel().validationMessages.clear();
+		getOrCreateComponentModel().validFileType = null;
+		getOrCreateComponentModel().validFileSize = null;
+	}
+
+	/**
+	 * Adds invalid file type validation message.
+	 * @param fieldValue True/False
+	 * @param message The message to add.
+	 * @param args The message arguments.
+	 */
+	private void addFileTypeValidation(final boolean fieldValue, final String message, final Object... args) {
+		getOrCreateComponentModel().validFileType = fieldValue;
+		if (!fieldValue) {
+			String invalidMessage = String.format(I18nUtilities.format(null, message), StringUtils.join(args, ","));
+			getOrCreateComponentModel().validationMessages.add(invalidMessage);
+		}
+	}
+
+	/**
+	 * Adds invalid file size validation message.
+	 * @param fieldValue True/False
+	 * @param message The message to add.
+	 * @param args The message arguments.
+	 */
+	private void addFileSizeValidation(final boolean fieldValue, final String message, final Object... args) {
+		getOrCreateComponentModel().validFileSize = fieldValue;
+		if (!fieldValue) {
+			String invalidMessage = String.format(I18nUtilities.format(null, message), args);
+			getOrCreateComponentModel().validationMessages.add(invalidMessage);
+		}
+	}
+
+	/**
+	 * Indicates whether the uploaded file is valid, if {@link getFileTypes()} is set.
+	 *
+	 * @return true if file type valid, false file type invalid, otherwise null.
+	 * @see {@link getFileValidationMessages()}
+	 */
+	public Boolean isFileTypeValid() {
+		return getComponentModel().validFileType;
+	}
+
```

Comment:
What do you think? If no file size is set, then is it a valid file size? Maybe this should return TRUE when no file size is set. Same with File type?

Reply:
When fileSize is not set it returns a null, same with fileTypes. What do you think?

Reply:
Returning null is always dangerous. I think it should be `boolean` and return `true` even if no file type set as technically it is valid. Same for `maxfilesize`.

Reply:
May I chime in with my opinions? Thanks... (feel free to disagree)

* By default ALL input is valid and accepted - we should be liberal in what we accept (is this the Robustness Principle?). 

  Constraints may be given but if they are not then there are no constraints *from a validation perspective*. There will be natural constraints but that is different to **validity**. 

  For example if we have a text input with no constraints then any input is valid. If we constrain it to only allow 20 characters with no whitespace then it is only valid if it meets both those constraints.

* Booleans should not be used to represent more than two states.
  I believe the API would be more intuitive if we used an explicit three states (e.g. an Enum).
  In this case it is purely academic because we only need two states.

Reply:
Yes agree @ricksbrown my last commit has fixed it, all Boolean are changed to boolean. Awaiting review.